### PR TITLE
quantile regression engines for xgboost and qrnn

### LINKED ISSUE #1321: more quantile engines
Related to #465 

The ones that come to mind are xgboost, lightgbm, and qrnn. The first and last are an easy update: 

``` r
library(parsnip)
library(bonsai)
library(qrnn)

lvls <- c(0.1, 0.5, 0.9)
```

``` r
# xgboost works out of the box

xgb_qnt <- xgb_train(
  x = mtcars[, -1],
  y = mtcars[[1]],
  quantile_alpha = lvls,
  objective = "reg:quantileerror"
)

xgb_predict(xgb_qnt, mtcars[1:2, -1])
#>          [,1]     [,2]     [,3]
#> [1,] 15.19863 21.00143 21.73931
#> [2,] 15.19863 21.00345 21.73931
```

``` r
# light gbm would need a wrapper since it only does one quantile level at a time

# fail:
train_lightgbm(
  x = mtcars[, -1],
  y = mtcars[[1]],
  alpha = lvls,
  objective = "quantile",
  quiet = TRUE
)
#> Error in `data$construct()`:
#> ! Parameter alpha should be of type double, got "0.1,0.5,0.9"

lgb_qnt <-
  train_lightgbm(
    x = mtcars[, -1],
    y = mtcars[[1]],
    alpha = lvls[2],
    objective = "quantile",
    quiet = TRUE
  )
```

``` r
# qrnn works out of the box

nn_qnt <- mcqrnn.fit(
  x = as.matrix(mtcars[, -1]),
  y = matrix(mtcars[[1]], ncol = 1),
  tau = lvls,
  n.hidden = 2,
  n.hidden2 = 2,
  n.trials = 1,
  iter.max = 500,
  trace = FALSE,
  method = "adam",
  alpha = 0.1,
  print.level = 0
)

mcqrnn.predict(as.matrix(mtcars[1:2, -1]), nn_qnt, tau = lvls)
#>       tau=0.1  tau=0.5  tau=0.9
#> [1,] 21.47180 21.59744 21.71403
#> [2,] 20.47568 20.64989 20.81849
```

<sup>Created on 2026-01-15 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

### PR DESCRIPTION
**Pull request by topepo**
Closes #1321 

Unit tests for the neural network engine will go in extratests

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -1,8 +1,10 @@
 # parsnip (development version)
 
+* xgboost and qrnn engines were added for censored regression to `boost_tree()` and `mlp()`, respectively. (#1321)
```
Comment (hfrick) [inline NEWS.md]:
```suggestion
* xgboost and qrnn engines were added for quantile regression to `boost_tree()` and `mlp()`, respectively. (#1321)
```

File: R/mlp_data.R [hunk 2]
```diff
@@ -679,3 +680,95 @@ set_pred(
...
+)
+
+
+set_fit(
+  model = "mlp",
+  eng = "qrnn",
+  mode = "quantile regression",
+  value = list(
+    interface = "matrix",
+    protect = c("x", "y"),
```
Comment (hfrick) [inline R/mlp_data.R:738]: The (parsnip) docs for this engine include case weights, should the weights be protected here?

File: man/rmd/mlp_qrnn.Rmd [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+This model has `r nrow(param)` tuning parameters:
+
+```{r}
+#| label: qrnn-param-list
+#| echo: false
+#| results: asis
+param$item
+```
+
+Other engine arguments of interest:
```
Comment (hfrick) [inline man/rmd/mlp_qrnn.Rmd:38]: These are not tunable, are they? Since they are listed in the section "Tuning Parameters", it might be good to clarify that they are not tunable.

File: R/mlp.R [hunk 4]
```diff
@@ -523,3 +523,50 @@ set_tf_seed <- function(seed) {
...
+
+#' Wrapper for `qrnn::mcqrnn.fit()`
+#' @export
+#' @param x Predictor matrix
+#' @param y outcome vector
+#' @param ... Options to pass to `qrnn::mcqrnn.fit()`
+#' @param trace Logical for printing
+#' @param tau Vector of quantile levels.
+#' @keywords internal
+mcqrnn_wrap <- function(x, y, ..., trace, tau) {
```
Comment (hfrick) [inline R/mlp.R]: Can we rename that to follow the naming convention for wrapper functions? The others are called `*_train()` , like `xgb_train()` and `ctree_train()`.

File: R/mlp.R [hunk 5]
```diff
@@ -523,3 +523,50 @@ set_tf_seed <- function(seed) {
...
+      if (inherits(act_fun, "try-error")) {
+        cli::cli_abort(
+          "Could not find an activation function called {.fn {dots$Th}} in the {.pkg qrnn} package."
+        )
+      }
+
+      act_deriv_fun <- try(
+        utils::getFromNamespace(deriv_nm, ns = "qrnn"),
+        silent = TRUE
+      )
```
Comment (hfrick) [inline R/mlp.R:558]: The activation function has got its own error message if the `try()` fails, let's also add one for this one here?

File: man/rmd/mlp_qrnn.Rmd (hunk 3 above)
Comment (topepo) [reply man/rmd/mlp_qrnn.Rmd:38]: I'll leave as-is and write a dials PR

File: R/mlp_data.R (hunk 2 above)
Comment (topepo) [reply R/mlp_data.R:738]: No, the wrong template was used.
